experimental/parser/parse_decl.go

```diff
@@ -0,0 +1,695 @@
+		}
+
+		semi, err := p.Punct(c, ";", in.After())
+		args.Semicolon = semi
+		if err != nil && args.ImportPath.Nil() {
+			p.Error(err)
+		}
+
+		return p.NewDeclImport(args).AsAny()
+
+	case "reserved", "extensions":
+		if next.Text() == "=" {
+			// If whatever follows the path is an =, we're going to assume this
+			// is trying to be a field.
+			break
+		}
+
+		// Otherwise, rewind the cursor to before we parsed a type, and
+		// parse a range instead. Rewinding is necessary because otherwise we get
+		// into an annoying situation where if we have e.g. reserved foo to bar;
+		// we have already consumed reserved foo, but we want to push foo
+		// through the expression machinery to get foo to bar as a single
+		// expression.
+		c.Rewind(mark)
+		return parseRange(p, c).AsAny()
+	}
+
+	return parseDef(p, c, kw, ty, path).AsAny()
+}
+
+// parseBody parses a ({}-delimited) body of declarations.
+func parseBody(p *parser, braces token.Token, in taxa.Subject) ast.DeclBody {
+	body := p.NewDeclBody(braces)
+
+	// Drain the contents of the body into it. Remember,
+	// parseDecl must always make progress if there is more to
+	// parse.
+	c := braces.Children()
+	for !c.Done() {
+		if next := parseDecl(p, c, in); !next.Nil() {
+			body.Append(next)
+		}
+	}
+
+	return body
+}
+
+// parseRange parses a reserved/extensions range.
+func parseRange(p *parser, c *token.Cursor) ast.DeclRange {
+	// Consume the keyword token.
+	kw := c.Pop()
+
+	in := taxa.Extensions
+	if kw.Text() == "reserved" {
+		in = taxa.Reserved
+	}
+
+	// Consume expressions until we hit a semicolon or the [ of a compact
+	// options. Note that this means that we do not parse reserved [1, 2, 3];
+	// "correctly".
+	var (
+		// badExpr keeps track of whether we exited the loop due to a parse
+		// error or because we hit ; or [ or EOF.
+		badExpr bool
+		exprs   []exprComma
+	)
+
+	// Parse expressions until we hit a semicolon or the [ of compact options.
+	elems := commas(p, c, true, func(c *token.Cursor) (ast.ExprAny, bool) {
+		next := c.Peek().Text()
+		if next == ";" || next == "[" {
+			return ast.ExprAny{}, false
+		}
+
+		expr := parseExpr(p, c, in.In())
+		badExpr = badExpr || expr.Nil()
+
+		return expr, !expr.Nil()
+	})
+	elems(func(expr ast.ExprAny, comma token.Token) bool {
+		exprs = append(exprs, exprComma{expr, comma})
+		return true
+	})
+
+	var options ast.CompactOptions
+	if next := c.Peek(); next.Text() == "[" && !next.IsLeaf() {
+		options = parseOptions(p, c.Pop(), in)
+	}
+
+	// Parse a semicolon, if possible.
+	semi, err := p.Punct(c, ";", in.After())
+	if err != nil && !badExpr {
+		p.Error(err)
+	}
+
+	r := p.NewDeclRange(ast.DeclRangeArgs{
+		Keyword:   kw,
+		Options:   options,
+		Semicolon: semi,
+	})
+	for _, e := range exprs {
+		r.AppendComma(e.expr, e.comma)
+	}
+
+	return r
+}
+
+// parseDef parses a generic definition.
+func parseDef(p *parser, c *token.Cursor, kw token.Token, ty ast.TypeAny, path ast.Path) ast.DeclDef {
```

Comment:
I refactored this to be more table-driven, and moved it into its own file.

Admittedly, this has no correlation with the EBNF you wrote; this function is how I was always planning to parse defs, and it's why the AST looks the way it does. I don't really expect this parser to be especially pretty, so the best solution, I think, is to add more comments with rationale wherever you think it's not obvious.

Reply:
Yeah, I understand why it doesn't correlate to the EBNF. But even aside form the flexible AST shape, there's a lot of logic for error productions -- intentionally looking for invalid constructions in order to diagnose them. I don't think it necessarily needs more comments for those. I suppose I was just lamenting some of that error handling and worried whether these diagnostics will really be worth the complication in logic.

Though, speaking of comments, one thing that would be good: maybe here in parseDef, or perhaps instead in the AST comments for DeclDef, to actually include the grammar productions from the spec for all of the things that get parsed via a Def and then also show the "effective" grammar for a Def. So having those next to each other in easier-to-read form makes it easier to comprehend how the Def can handle parsing of all of those different parts of the language, and then there's a production defined for the Def itself that the code can be compared against. WDYT? (Doesn't have to be this PR -- could be in a follow-up.)

Reply:
That last suggestion is a great idea. Here's what I think we should do: every struct in the ast package should contain an EBNF of what it can represent. I've done this before with some other parsers I've written and it's useful.
I can work on this in parallel to this PR.